SpatialGDK/Source/SpatialGDK/Private/Utils/LockingStatics.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Utils/LockingStatics.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "LoadBalancing/AbstractLockingPolicy.h"
+#include "SpatialConstants.h"
+#include "Utils/SpatialStatics.h"
+
+#include "Containers/UnrealString.h"
+
+DEFINE_LOG_CATEGORY(LogLocking);
+
+namespace
+{
+bool CanProcessActor(const AActor* Actor)
+{
+	if (Actor == nullptr)
+	{
+		UE_LOG(LogLocking, Error, TEXT("Calling locking API functions on nullptr Actor is invalid."));
+		return false;
+	}
+
+	const UNetDriver* NetDriver = Actor->GetWorld()->GetNetDriver();
+	if (!NetDriver->IsServer())
+	{
+		UE_LOG(LogLocking, Error, TEXT("Calling locking API functions on a client is invalid. Actor: %s"), *GetNameSafe(Actor));
+		return false;
+	}
+
+	if (!Actor->HasAuthority())
+	{
+		UE_LOG(LogLocking, Error, TEXT("Calling locking API functions on a non-auth Actor is invalid. Actor: %s."),
+			*GetNameSafe(Actor));
+		return false;
+	}
+
+	return true;
+}
+} // anonymous namespace
+
+FLockingToken ULockingStatics::AcquireLock(AActor* Actor)
+{
+	if (!CanProcessActor(Actor) || !USpatialStatics::IsSpatialMultiWorkerEnabled(Actor->GetWorld()))
+	{
+		return FLockingToken{ SpatialConstants::INVALID_ACTOR_LOCK_TOKEN };
+	}
+
+	UAbstractLockingPolicy* LockingPolicy = Cast<USpatialNetDriver>(Actor->GetWorld()->GetNetDriver())->LockingPolicy;
+
+	const uint32 NewLockCount = LockingPolicy->GetActorLockCount(Actor) + 1;
+
+	const ActorLockToken LockToken = LockingPolicy->AcquireLock(Actor,
+		FString::Printf(TEXT("Actor %s locked. Now locked %d times."), *Actor->GetName(), NewLockCount));
+
+	UE_LOG(LogLocking, Verbose, TEXT("LockingComponent called AcquireLock. Actor: %s. Token: %lld. New lock count: %d"),
+		*Actor->GetName(), LockToken, NewLockCount);
+
+	return FLockingToken{ LockToken };
+}
+
+bool ULockingStatics::IsLocked(const AActor* Actor)
+{
+	if (!CanProcessActor(Actor))
+	{
+		return false;
+	}
+
+	// If multi worker is disabled, we want the server to behave as if it always has authority, and so we say the Actor is locked.
```

Comment:
Why locked instead of unlocked or just leaving the locking system in place and happening to never try to migrate?

Reply:
I don't really have strong feelings on what it should be, but I'm curious as to why this choice.

Reply:
Actors will never try to migrate or anything but it's still an API call users can make from game code. I see being in a locked state as a guaranteed of authority, which we have in single-server / native, so opted for that.

Reply:
I'm not sure I agree - I think authority should be used as the guarantee of authority - and if you attempt to lock with multiworker disabled we either allow the users to lock their actors, or we return false for everything.

Reply:
will switch to return false 👍